l20n/migrate/transforms.py

```diff
@@ -0,0 +1,247 @@
+        keys_in_order = sorted(
+            replacements.keys(),
+            lambda x, y: self.source.find(x) - self.source.find(y)
+        )
+
+        # Used to reduce the `keys_in_order` list.
+        def replace(acc, cur):
+            """Convert original placeables and text into FTL Nodes.
+
+            For each original placeable the translation will be partitioned
+            around it and the text before it will be converted into an
+            `FTL.TextElement` and the placeable will be converted to an
+            `FTL.Placeable`. The text following the placebale will be fed again
+            to the `replace` function.
+            """
+
+            parts, rest = acc
+            before, key, after = rest.value.partition(cur)
+
+            placeable = FTL.Placeable(replacements[key])
+
+            # Return the elements found and converted so far, and the remaining
+            # text which hasn't been scanned for placeables yet.
+            return (
+                parts + [FTL.TextElement(before), placeable],
+                FTL.TextElement(after)
+            )
+
+        def is_non_empty(elem):
+            """Used for filtering empty `FTL.TextElement` nodes out."""
+            return not isinstance(elem, FTL.TextElement) or len(elem.value)
+
+        # Start with an empty list of elements and the original translation.
+        init = ([], FTL.TextElement(self.source))
+        parts, tail = reduce(replace, keys_in_order, init)
+
+        # Explicitly concat the trailing part to get the full list of elements
+        # and filter out the empty ones.
+        elements = filter(is_non_empty, parts + [tail])
+        quoted = self.source.startswith(' ') or self.source.endswith(' ')
+
+        return FTL.Pattern(None, elements, quoted)
+
+
+class PLURALS(Transform):
+    """Convert semicolon-separated variants into a select expression.
+
+    Build an `FTL.SelectExpression` with the supplied `selector` and variants
+    extracted from the `source`.  Each variant will be run through the
+    `foreach` function.  It should return an `FTL.Pattern`.
+    """
+
+    def __init__(self, source, selector, foreach):
+        self.source = source
+        self.selector = selector
+        self.foreach = foreach
+
+    def __call__(self, ctx):
+        variants = self.source.split(';')
+        keys_enum = enumerate(ctx.plural_categories)
+        last_variant_index = len(variants) - 1
+
+        def createMember(variant):
+            index, key = next(keys_enum)
+            # Run the legacy variant through `foreach` which returns an
+            # `FTL.Node` describing the transformation required for each
+            # variant.  Then evaluate it to a migrated FTL node.
+            value = evaluate(ctx, self.foreach(variant))
+            return FTL.Member(
+                FTL.Keyword(key),
+                value,
+                default=index == last_variant_index
+            )
```

Comment:
There are two issues here, thanks for pointing them out.

1. There's a bug here:  if there are more variants than CLDR plural categories, line 165 will raise a `StopIteration` exception. I think using something similar to `zip` will help.

2. Right now this code will set as default the last variant taken from the legacy translation.  After I fix #1 maybe we should just always set as default the last variant in the resulting select expression?